From 5dc062a8b947d9c131b74a615f5472037257d6c8 Mon Sep 17 00:00:00 2001
From: charles <dminus@andestech.com>
Date: Thu, 18 Jan 2024 18:20:06 +0800
Subject: [PATCH 027/168] cache: andes: Replace specific CPU name to a more
 generic name

The ax45mp_cache.c file is not only usable with the ax45mp CPU but
can also be used on other CPU types within the Andes platform.

Signed-off-by: charles <dminus@andestech.com>
---
 arch/riscv/Kconfig.socs      |   2 +-
 drivers/cache/Kconfig        |   6 +-
 drivers/cache/Makefile       |   2 +-
 drivers/cache/andes_cache.c  | 232 ++++++++++++++++++++++++++++++++++
 drivers/cache/ax45mp_cache.c | 236 -----------------------------------
 drivers/soc/renesas/Kconfig  |   2 +-
 6 files changed, 238 insertions(+), 242 deletions(-)
 create mode 100644 drivers/cache/andes_cache.c
 delete mode 100644 drivers/cache/ax45mp_cache.c

diff --git a/arch/riscv/Kconfig.socs b/arch/riscv/Kconfig.socs
index cb44df4cb5ea..5052dd79bba7 100644
--- a/arch/riscv/Kconfig.socs
+++ b/arch/riscv/Kconfig.socs
@@ -5,7 +5,7 @@ config ARCH_ANDES
 	depends on MMU && !XIP_KERNEL
 	select ERRATA_ANDES
 	select ERRATA_ANDES_CMO
-	select AX45MP_L2_CACHE
+	select ANDES_CACHE
 	help
 	  This enables support for Andes platform hardware.
 
diff --git a/drivers/cache/Kconfig b/drivers/cache/Kconfig
index d6e5e3abaad8..10a852ddda90 100644
--- a/drivers/cache/Kconfig
+++ b/drivers/cache/Kconfig
@@ -1,11 +1,11 @@
 # SPDX-License-Identifier: GPL-2.0
 menu "Cache Drivers"
 
-config AX45MP_L2_CACHE
-	bool "Andes Technology AX45MP L2 Cache controller"
+config ANDES_CACHE
+	bool "Andes platform CPUs Cache controller"
 	depends on RISCV
 	select RISCV_NONSTANDARD_CACHE_OPS
 	help
-	  Support for the L2 cache controller on Andes Technology AX45MP platforms.
+	  Support for the L1 and L2 cache controller on Andes platform CPUs.
 
 endmenu
diff --git a/drivers/cache/Makefile b/drivers/cache/Makefile
index 2012e7fb978d..e2830f9f35b2 100644
--- a/drivers/cache/Makefile
+++ b/drivers/cache/Makefile
@@ -1,3 +1,3 @@
 # SPDX-License-Identifier: GPL-2.0
 
-obj-$(CONFIG_AX45MP_L2_CACHE) += ax45mp_cache.o
+obj-$(CONFIG_ANDES_CACHE) += andes_cache.o
diff --git a/drivers/cache/andes_cache.c b/drivers/cache/andes_cache.c
new file mode 100644
index 000000000000..b068a4990543
--- /dev/null
+++ b/drivers/cache/andes_cache.c
@@ -0,0 +1,232 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * non-coherent cache operations for Andes Platform CPUs.
+ *
+ * Copyright (C) 2023 Renesas Electronics Corp.
+ */
+
+#include <linux/cacheflush.h>
+#include <linux/cacheinfo.h>
+#include <linux/dma-direction.h>
+#include <linux/of_address.h>
+#include <linux/of_platform.h>
+
+#include <asm/dma-noncoherent.h>
+
+/* L2 cache registers */
+#define ANDES_L2C_REG_CTL_OFFSET		0x8
+
+#define ANDES_L2C_REG_C0_CMD_OFFSET		0x40
+#define ANDES_L2C_REG_C0_ACC_OFFSET		0x48
+#define ANDES_L2C_REG_STATUS_OFFSET		0x80
+
+/* D-cache operation */
+#define ANDES_CCTL_L1D_VA_INVAL			0 /* Invalidate an L1 cache entry */
+#define ANDES_CCTL_L1D_VA_WB			1 /* Write-back an L1 cache entry */
+
+/* L2 CCTL status */
+#define ANDES_CCTL_L2_STATUS_IDLE		0
+
+/* L2 CCTL status cores mask */
+#define ANDES_CCTL_L2_STATUS_C0_MASK		0xf
+
+/* L2 cache operation */
+#define ANDES_CCTL_L2_PA_INVAL			0x8 /* Invalidate an L2 cache entry */
+#define ANDES_CCTL_L2_PA_WB			0x9 /* Write-back an L2 cache entry */
+
+#define ANDES_L2C_REG_PER_CORE_OFFSET		0x10
+#define ANDES_CCTL_L2_STATUS_PER_CORE_OFFSET	4
+
+#define ANDES_L2C_REG_CN_CMD_OFFSET(n)	\
+	(ANDES_L2C_REG_C0_CMD_OFFSET + ((n) * ANDES_L2C_REG_PER_CORE_OFFSET))
+#define ANDES_L2C_REG_CN_ACC_OFFSET(n)	\
+	(ANDES_L2C_REG_C0_ACC_OFFSET + ((n) * ANDES_L2C_REG_PER_CORE_OFFSET))
+#define ANDES_CCTL_L2_STATUS_CN_MASK(n)	\
+	(ANDES_CCTL_L2_STATUS_C0_MASK << ((n) * ANDES_CCTL_L2_STATUS_PER_CORE_OFFSET))
+
+#define ANDES_CCTL_REG_UCCTLBEGINADDR_NUM	0x80b
+#define ANDES_CCTL_REG_UCCTLCOMMAND_NUM		0x80c
+
+struct andes_priv {
+	void __iomem *l2c_base;
+	u32 andes_cache_line_size;
+};
+
+static struct andes_priv andes_priv;
+
+/* L2 Cache operations */
+static inline uint32_t andes_cpu_l2c_get_cctl_status(void)
+{
+	return readl(andes_priv.l2c_base + ANDES_L2C_REG_STATUS_OFFSET);
+}
+
+static void andes_cpu_cache_operation(unsigned long start, unsigned long end,
+				       unsigned int l1_op, unsigned int l2_op)
+{
+	unsigned long line_size = andes_priv.andes_cache_line_size;
+	void __iomem *base = andes_priv.l2c_base;
+	int mhartid = smp_processor_id();
+	unsigned long pa;
+
+	if (base) {
+		while (end > start) {
+			csr_write(ANDES_CCTL_REG_UCCTLBEGINADDR_NUM, start);
+			csr_write(ANDES_CCTL_REG_UCCTLCOMMAND_NUM, l1_op);
+
+			pa = virt_to_phys((void *)start);
+			writel(pa, base + ANDES_L2C_REG_CN_ACC_OFFSET(mhartid));
+			writel(l2_op, base + ANDES_L2C_REG_CN_CMD_OFFSET(mhartid));
+			while ((andes_cpu_l2c_get_cctl_status() &
+				ANDES_CCTL_L2_STATUS_CN_MASK(mhartid)) !=
+				ANDES_CCTL_L2_STATUS_IDLE)
+				;
+
+			start += line_size;
+		}
+	} else {
+		while (end > start) {
+			csr_write(ANDES_CCTL_REG_UCCTLBEGINADDR_NUM, start);
+			csr_write(ANDES_CCTL_REG_UCCTLCOMMAND_NUM, l1_op);
+
+			start += line_size;
+		}
+	}
+}
+
+/* Write-back L1 and L2 cache entry */
+static inline void andes_cpu_dcache_wb_range(unsigned long start, unsigned long end)
+{
+	andes_cpu_cache_operation(start, end, ANDES_CCTL_L1D_VA_WB,
+				   ANDES_CCTL_L2_PA_WB);
+}
+
+/* Invalidate the L1 and L2 cache entry */
+static inline void andes_cpu_dcache_inval_range(unsigned long start, unsigned long end)
+{
+	andes_cpu_cache_operation(start, end, ANDES_CCTL_L1D_VA_INVAL,
+				   ANDES_CCTL_L2_PA_INVAL);
+}
+
+static void andes_dma_cache_inv(phys_addr_t paddr, size_t size)
+{
+	unsigned long start = (unsigned long)phys_to_virt(paddr);
+	unsigned long end = start + size;
+	unsigned long line_size;
+	unsigned long flags;
+
+	if (unlikely(start == end))
+		return;
+
+	line_size = andes_priv.andes_cache_line_size;
+
+	start = start & (~(line_size - 1));
+	end = ((end + line_size - 1) & (~(line_size - 1)));
+
+	local_irq_save(flags);
+
+	andes_cpu_dcache_inval_range(start, end);
+
+	local_irq_restore(flags);
+}
+
+static void andes_dma_cache_wback(phys_addr_t paddr, size_t size)
+{
+	unsigned long start = (unsigned long)phys_to_virt(paddr);
+	unsigned long end = start + size;
+	unsigned long line_size;
+	unsigned long flags;
+
+	line_size = andes_priv.andes_cache_line_size;
+	start = start & (~(line_size - 1));
+	local_irq_save(flags);
+	andes_cpu_dcache_wb_range(start, end);
+	local_irq_restore(flags);
+}
+
+static void andes_dma_cache_wback_inv(phys_addr_t paddr, size_t size)
+{
+	andes_dma_cache_wback(paddr, size);
+	andes_dma_cache_inv(paddr, size);
+}
+
+static int andes_get_l2_line_size(struct device_node *np)
+{
+	int ret;
+
+	ret = of_property_read_u32(np, "cache-line-size", &andes_priv.andes_cache_line_size);
+	if (ret) {
+		pr_err("Failed to get cache-line-size.\n");
+		return ret;
+	}
+
+	return 0;
+}
+
+static const struct riscv_nonstd_cache_ops andes_cmo_ops __initconst = {
+	.wback = &andes_dma_cache_wback,
+	.inv = &andes_dma_cache_inv,
+	.wback_inv = &andes_dma_cache_wback_inv,
+};
+
+static const struct of_device_id andes_cache_ids[] = {
+	{ .compatible = "andestech,ax45mp-cache" },
+	{ .compatible = "cache" },
+	{ /* sentinel */ }
+};
+
+static int __init andes_cache_init(void)
+{
+	struct device_node *np;
+	struct resource res;
+	int ret;
+
+	/*
+	 * Initialize l2c_base and cache_line_size to provide
+	 * default settings in the absence of an l2c layout.
+	 */
+	andes_priv.l2c_base = 0;
+
+	np = of_cpu_device_node_get(0);
+	if (!of_device_is_available(np))
+		return -ENODEV;
+
+	ret = of_property_read_u32(np, "d-cache-line-size",
+				   &andes_priv.andes_cache_line_size);
+	if (ret) {
+		pr_err("Failed to get d-cache-line-size.\n");
+		return ret;
+	}
+
+	riscv_noncoherent_register_cache_ops(&andes_cmo_ops);
+
+	np = of_find_matching_node(NULL, andes_cache_ids);
+	if (!of_device_is_available(np))
+		return -ENODEV;
+
+	ret = of_address_to_resource(np, 0, &res);
+	if (ret)
+		return ret;
+
+	/*
+	 * If IOCP is present on the Andes ANDES core riscv_cbom_block_size
+	 * will be 0 for sure, so we can definitely rely on it. If
+	 * riscv_cbom_block_size = 0 we don't need to handle CMO using SW any
+	 * more so we just return success here and only if its being set we
+	 * continue further in the probe path.
+	 */
+	if (!riscv_cbom_block_size)
+		return 0;
+
+	andes_priv.l2c_base = ioremap(res.start, resource_size(&res));
+	if (!andes_priv.l2c_base)
+		return -ENOMEM;
+
+	ret = andes_get_l2_line_size(np);
+	if (ret) {
+		iounmap(andes_priv.l2c_base);
+		return ret;
+	}
+
+	return 0;
+}
+early_initcall(andes_cache_init);
diff --git a/drivers/cache/ax45mp_cache.c b/drivers/cache/ax45mp_cache.c
deleted file mode 100644
index 8c84627219c9..000000000000
--- a/drivers/cache/ax45mp_cache.c
+++ /dev/null
@@ -1,236 +0,0 @@
-// SPDX-License-Identifier: GPL-2.0
-/*
- * non-coherent cache functions for Andes AX45MP
- *
- * Copyright (C) 2023 Renesas Electronics Corp.
- */
-
-#include <linux/cacheflush.h>
-#include <linux/cacheinfo.h>
-#include <linux/dma-direction.h>
-#include <linux/of_address.h>
-#include <linux/of_platform.h>
-
-#include <asm/dma-noncoherent.h>
-
-/* L2 cache registers */
-#define AX45MP_L2C_REG_CTL_OFFSET		0x8
-
-#define AX45MP_L2C_REG_C0_CMD_OFFSET		0x40
-#define AX45MP_L2C_REG_C0_ACC_OFFSET		0x48
-#define AX45MP_L2C_REG_STATUS_OFFSET		0x80
-
-/* D-cache operation */
-#define AX45MP_CCTL_L1D_VA_INVAL		0 /* Invalidate an L1 cache entry */
-#define AX45MP_CCTL_L1D_VA_WB			1 /* Write-back an L1 cache entry */
-
-/* L2 CCTL status */
-#define AX45MP_CCTL_L2_STATUS_IDLE		0
-
-/* L2 CCTL status cores mask */
-#define AX45MP_CCTL_L2_STATUS_C0_MASK		0xf
-
-/* L2 cache operation */
-#define AX45MP_CCTL_L2_PA_INVAL			0x8 /* Invalidate an L2 cache entry */
-#define AX45MP_CCTL_L2_PA_WB			0x9 /* Write-back an L2 cache entry */
-
-#define AX45MP_L2C_REG_PER_CORE_OFFSET		0x10
-#define AX45MP_CCTL_L2_STATUS_PER_CORE_OFFSET	4
-
-#define AX45MP_L2C_REG_CN_CMD_OFFSET(n)	\
-	(AX45MP_L2C_REG_C0_CMD_OFFSET + ((n) * AX45MP_L2C_REG_PER_CORE_OFFSET))
-#define AX45MP_L2C_REG_CN_ACC_OFFSET(n)	\
-	(AX45MP_L2C_REG_C0_ACC_OFFSET + ((n) * AX45MP_L2C_REG_PER_CORE_OFFSET))
-#define AX45MP_CCTL_L2_STATUS_CN_MASK(n)	\
-	(AX45MP_CCTL_L2_STATUS_C0_MASK << ((n) * AX45MP_CCTL_L2_STATUS_PER_CORE_OFFSET))
-
-#define AX45MP_CCTL_REG_UCCTLBEGINADDR_NUM	0x80b
-#define AX45MP_CCTL_REG_UCCTLCOMMAND_NUM	0x80c
-
-struct ax45mp_priv {
-	void __iomem *l2c_base;
-	u32 ax45mp_cache_line_size;
-};
-
-static struct ax45mp_priv ax45mp_priv;
-
-/* L2 Cache operations */
-static inline uint32_t ax45mp_cpu_l2c_get_cctl_status(void)
-{
-	return readl(ax45mp_priv.l2c_base + AX45MP_L2C_REG_STATUS_OFFSET);
-}
-
-static void ax45mp_cpu_cache_operation(unsigned long start, unsigned long end,
-				       unsigned int l1_op, unsigned int l2_op)
-{
-	unsigned long line_size = ax45mp_priv.ax45mp_cache_line_size;
-	void __iomem *base = ax45mp_priv.l2c_base;
-	int mhartid = smp_processor_id();
-	unsigned long pa;
-
-	if (base) {
-		while (end > start) {
-			csr_write(AX45MP_CCTL_REG_UCCTLBEGINADDR_NUM, start);
-			csr_write(AX45MP_CCTL_REG_UCCTLCOMMAND_NUM, l1_op);
-
-			pa = virt_to_phys((void *)start);
-			writel(pa, base + AX45MP_L2C_REG_CN_ACC_OFFSET(mhartid));
-			writel(l2_op, base + AX45MP_L2C_REG_CN_CMD_OFFSET(mhartid));
-			while ((ax45mp_cpu_l2c_get_cctl_status() &
-				AX45MP_CCTL_L2_STATUS_CN_MASK(mhartid)) !=
-				AX45MP_CCTL_L2_STATUS_IDLE)
-				;
-
-			start += line_size;
-		}
-	} else {
-		while (end > start) {
-			csr_write(AX45MP_CCTL_REG_UCCTLBEGINADDR_NUM, start);
-			csr_write(AX45MP_CCTL_REG_UCCTLCOMMAND_NUM, l1_op);
-
-			start += line_size;
-		}
-	}
-}
-
-/* Write-back L1 and L2 cache entry */
-static inline void ax45mp_cpu_dcache_wb_range(unsigned long start, unsigned long end)
-{
-	ax45mp_cpu_cache_operation(start, end, AX45MP_CCTL_L1D_VA_WB,
-				   AX45MP_CCTL_L2_PA_WB);
-}
-
-/* Invalidate the L1 and L2 cache entry */
-static inline void ax45mp_cpu_dcache_inval_range(unsigned long start, unsigned long end)
-{
-	ax45mp_cpu_cache_operation(start, end, AX45MP_CCTL_L1D_VA_INVAL,
-				   AX45MP_CCTL_L2_PA_INVAL);
-}
-
-static void ax45mp_dma_cache_inv(phys_addr_t paddr, size_t size)
-{
-	unsigned long start = (unsigned long)phys_to_virt(paddr);
-	unsigned long end = start + size;
-	unsigned long line_size;
-	unsigned long flags;
-
-	if (unlikely(start == end))
-		return;
-
-	line_size = ax45mp_priv.ax45mp_cache_line_size;
-
-	start = start & (~(line_size - 1));
-	end = ((end + line_size - 1) & (~(line_size - 1)));
-
-	local_irq_save(flags);
-
-	ax45mp_cpu_dcache_inval_range(start, end);
-
-	local_irq_restore(flags);
-}
-
-static void ax45mp_dma_cache_wback(phys_addr_t paddr, size_t size)
-{
-	unsigned long start = (unsigned long)phys_to_virt(paddr);
-	unsigned long end = start + size;
-	unsigned long line_size;
-	unsigned long flags;
-
-	if (unlikely(start == end))
-		return;
-
-	line_size = ax45mp_priv.ax45mp_cache_line_size;
-	start = start & (~(line_size - 1));
-	end = ((end + line_size - 1) & (~(line_size - 1)));
-	local_irq_save(flags);
-	ax45mp_cpu_dcache_wb_range(start, end);
-	local_irq_restore(flags);
-}
-
-static void ax45mp_dma_cache_wback_inv(phys_addr_t paddr, size_t size)
-{
-	ax45mp_dma_cache_wback(paddr, size);
-	ax45mp_dma_cache_inv(paddr, size);
-}
-
-static int ax45mp_get_l2_line_size(struct device_node *np)
-{
-	int ret;
-
-	ret = of_property_read_u32(np, "cache-line-size", &ax45mp_priv.ax45mp_cache_line_size);
-	if (ret) {
-		pr_err("Failed to get cache-line-size.\n");
-		return ret;
-	}
-
-	return 0;
-}
-
-static const struct riscv_nonstd_cache_ops ax45mp_cmo_ops __initdata = {
-	.wback = &ax45mp_dma_cache_wback,
-	.inv = &ax45mp_dma_cache_inv,
-	.wback_inv = &ax45mp_dma_cache_wback_inv,
-};
-
-static const struct of_device_id ax45mp_cache_ids[] = {
-	{ .compatible = "andestech,ax45mp-cache" },
-	{ .compatible = "cache" },
-	{ /* sentinel */ }
-};
-
-static int __init ax45mp_cache_init(void)
-{
-	struct device_node *np;
-	struct resource res;
-	int ret;
-
-	/*
-	 * Initialize l2c_base and cache_line_size to provide
-	 * default settings in the absence of an l2c layout.
-	 */
-	ax45mp_priv.l2c_base = 0;
-
-	np = of_cpu_device_node_get(0);
-	if (!of_device_is_available(np))
-		return -ENODEV;
-
-	ret = of_property_read_u32(np, "d-cache-line-size",
-				   &ax45mp_priv.ax45mp_cache_line_size);
-	if (ret) {
-		pr_err("Failed to get d-cache-line-size.\n");
-		return ret;
-	}
-
-	riscv_noncoherent_register_cache_ops(&ax45mp_cmo_ops);
-
-	np = of_find_matching_node(NULL, ax45mp_cache_ids);
-	if (!of_device_is_available(np))
-		return -ENODEV;
-
-	ret = of_address_to_resource(np, 0, &res);
-	if (ret)
-		return ret;
-
-	/*
-	 * If IOCP is present on the Andes AX45MP core riscv_cbom_block_size
-	 * will be 0 for sure, so we can definitely rely on it. If
-	 * riscv_cbom_block_size = 0 we don't need to handle CMO using SW any
-	 * more so we just return success here and only if its being set we
-	 * continue further in the probe path.
-	 */
-	if (!riscv_cbom_block_size)
-		return 0;
-
-	ax45mp_priv.l2c_base = ioremap(res.start, resource_size(&res));
-	if (!ax45mp_priv.l2c_base)
-		return -ENOMEM;
-
-	ret = ax45mp_get_l2_line_size(np);
-	if (ret) {
-		iounmap(ax45mp_priv.l2c_base);
-		return ret;
-	}
-
-	return 0;
-}
-early_initcall(ax45mp_cache_init);
diff --git a/drivers/soc/renesas/Kconfig b/drivers/soc/renesas/Kconfig
index acc812e490d0..bc5b470ab46d 100644
--- a/drivers/soc/renesas/Kconfig
+++ b/drivers/soc/renesas/Kconfig
@@ -338,7 +338,7 @@ config ARCH_R9A07G043
 	depends on !RISCV_ISA_ZICBOM
 	depends on RISCV_SBI
 	select ARCH_RZG2L
-	select AX45MP_L2_CACHE
+	select ANDES_CACHE
 	select DMA_GLOBAL_POOL
 	select ERRATA_ANDES
 	select ERRATA_ANDES_CMO
-- 
2.34.1

